libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -614,73 +645,73 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		ArrayType const& arrayType = dynamic_cast<ArrayType const&>(*_functionCall.annotation().type);
 		solAssert(arguments.size() == 1, "");
 
-		defineExpression(_functionCall) <<
+		IRVariable value = convert(*arguments[0], *TypeProvider::uint256());
+		define(_functionCall) <<
 			m_utils.allocateMemoryArrayFunction(arrayType) <<
 			"(" <<
-				expressionAsType(*arguments[0], *TypeProvider::uint256()) <<
+			value.commaSeparatedList() <<
 			")\n";
-
 		break;
 	}
 	case FunctionType::Kind::KECCAK256:
 	{
 		solAssert(arguments.size() == 1, "");
 
 		ArrayType const* arrayType = TypeProvider::bytesMemory();
-		string const& array = m_context.newYulVariable();
-		m_code << "let " << array << " := " << expressionAsType(*arguments[0], *arrayType) << "\n";
+		auto array = convert(*arguments[0], *arrayType);
 
-		defineExpression(_functionCall) <<
+		define(_functionCall) <<
 			"keccak256(" <<
-			m_utils.arrayDataAreaFunction(*arrayType) << "(" <<
-			array <<
+			m_utils.arrayDataAreaFunction(*arrayType) + "(" <<
+			array.commaSeparatedList() <<
 			"), " <<
 			m_utils.arrayLengthFunction(*arrayType) <<
 			"(" <<
-			array <<
+			array.commaSeparatedList() <<
 			"))\n";
-
-			break;
+		break;
 	}
 	case FunctionType::Kind::ArrayPop:
 	{
-		ArrayType const& arrayType = dynamic_cast<ArrayType const&>(
-			*dynamic_cast<MemberAccess const&>(_functionCall.expression()).expression().annotation().type
-		);
-		defineExpression(_functionCall) <<
+		auto const& memberAccessExpression = dynamic_cast<MemberAccess const&>(_functionCall.expression()).expression();
+		ArrayType const& arrayType = dynamic_cast<ArrayType const&>(*memberAccessExpression.annotation().type);
+		define(_functionCall) <<
 			m_utils.storageArrayPopFunction(arrayType) <<
 			"(" <<
-			m_context.variable(_functionCall.expression()) <<
+			IRVariable(memberAccessExpression).commaSeparatedList() <<
 			")\n";
 		break;
 	}
 	case FunctionType::Kind::ArrayPush:
 	{
-		ArrayType const& arrayType = dynamic_cast<ArrayType const&>(
-			*dynamic_cast<MemberAccess const&>(_functionCall.expression()).expression().annotation().type
-		);
+		auto const& memberAccessExpression = dynamic_cast<MemberAccess const&>(_functionCall.expression()).expression();
+		ArrayType const& arrayType = dynamic_cast<ArrayType const&>(*memberAccessExpression.annotation().type);
 		if (arguments.empty())
 		{
 			auto slotName = m_context.newYulVariable();
 			auto offsetName = m_context.newYulVariable();
 			m_code << "let " << slotName << ", " << offsetName << " := " <<
```

Comment:
So storage lvalues are not covered by IRVariables?

Reply:
The problem is that value types don't know that they live in storage... so if ``arrayType.baseType()`` is e.g. ``uint256``, then I'd end up with an ``IRVariable`` of that type referring to a simple integer on stack instead of a slot and an offset in storage *pointing to* the integer...
That's actually the only reason for keeping the ``lvalue`` stuff at all - if we had a dedicated types for value types in storage and memory, then we could improve this, but those would not necessarily be a good idea in general and would be a huge effort to implement... but without something like that I haven't come up with a nice solution...

Unless we go one step further and define acutal ``IRType``s to go together with ``IRVariable``s :-D...